Fix None and equal_to_1 arguments issue in Triton kernel generated by AOTI

renganxu opened the pull request:
Summary:
When a Triton kernel has arguments with None values followed by arguments with value 1, AOTI attempts to remove the None arguments and update the indices of the equal_to_1 arguments in triton_meta["configs"]. However, if the same kernel is called multiple times, this optimization process is repeated. Prior to this diff, the indices of equal_to_1 arguments from subsequent calls (second and later) were based on the updated indices from the previous call, resulting in incorrect behavior.
This diff aims to localize the updated indices for equal_to_1 arguments within the optimization process of the current call, ensuring accurate and consistent results.

Test Plan:
Unit Test:
```
buck2 run mode/dev-nosan caffe2/test/inductor:test_aot_inductor -- -r test_triton_kernel_with_none_inputs_and_equal_to_1_arg
```

Differential Revision: D69998314




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

renganxu commented:
@pytorchbot label "topic: not user facing"

ColinPeppler commented on torch/_inductor/codegen/cpp_wrapper_gpu.py:
```diff
@@ -509,6 +512,7 @@ def generate_kernel_call(
         grid_fn: str = "grid",
         triton_meta=None,
         autotune_configs=None,
+        equal_to_1: list[int]=None,
```
if you add this argument to `generate_kernel_call` for cpp_wrapper_gpu, you'd need to do it for "PythonWrapperCodegen" and "CppWrapperCpu".

is there a way to do this without adding `equal_to_1` arg here?

renganxu replied:
Thanks. The lint also suggested this change. In the latest version, I followed @davidberard98 's advice to use a deepcopy for triton_meta. Now it does not need to pass equal_to_1 arg anymore.

davidberard98 commented on torch/_inductor/ir.py:
```diff
@@ -5817,6 +5817,7 @@ def codegen(self, wrapper) -> None:  # type: ignore[no-untyped-def]
             if kernel.arg_names.index(kwarg) in kernel.constexprs:
                 constexpr_indices.append(idx)
 
+        equal_to_1 = triton_meta["configs"][0].equal_to_1
```
note: we need to remove this. In future triton versions (i.e. branch cut  - end of next week) triton_meta["configs"][0].equal_to_1 will not exist)

davidberard98 commented on torch/_inductor/codegen/cpp_wrapper_gpu.py:
```diff
@@ -596,7 +600,8 @@ def generate_kernel_call(
                 ):
-                    equal_to_1 = triton_meta["configs"][0].equal_to_1
+                    if equal_to_1 is None:
+                        equal_to_1 = triton_meta["configs"][0].equal_to_1
```
also here, need to check triton_version_uses_attrs_dict to see whether equals_to_1 is a valid thing to be checking on the config

davidberard98 reviewed (commented):
thanks for taking a look at this! I have a few comments and I plan to come back once my local build of your branch finishes, but I'm leaving this (partially as notes to myself) in case I get distracted before the build finishes

davidberard98 reviewed (commented):
wdyt about just adding a `triton_meta = copy.deepcopy(triton_meta)` after line 5807 to avoid modifying the cached value of triton_meta?

renganxu commented:
> wdyt about just adding a `triton_meta = copy.deepcopy(triton_meta)` after line 5807 to avoid modifying the cached value of triton_meta?

Thanks for the suggestion! This looks like a better solution, so that we do not need to pass equal_to_1 and modify other interfaces. Let me try that

renganxu committed:
Fix None and equal_to_1 arguments issue in Triton kernel generated by AOTI (#148102)

Summary:

When a Triton kernel has arguments with None values followed by arguments with value 1, AOTI attempts to remove the None arguments and update the indices of the equal_to_1 arguments in triton_meta["configs"]. However, if the same kernel is called multiple times, this optimization process is repeated. Prior to this diff, the indices of equal_to_1 arguments from subsequent calls (second and later) were based on the updated indices from the previous call, resulting in incorrect behavior.
e.g. assume a kernel has the following list of arguments:
``` 
tensor_a, None, None, tensor_b, 1, tensor_c  
```
the index of the argument 1 is 4. After removing None arguments, the index of argument 1 is updated to 2

Prior to the fix, if the same kernel was called and processed again, an issue arose from the second call. Specifically, in the 2nd call, after removing None arguments, the index of argument 1 was incorrectly updated to 0. The index was also incorrect in the following calls.

This diff aims to localize the updated indices for equal_to_1 arguments within the optimization process of the current call, ensuring accurate and consistent results.

Test Plan:
Unit Test:
```
buck2 run mode/dev-nosan caffe2/test/inductor:test_aot_inductor -- -r test_triton_kernel_with_none_inputs_and_equal_to_1_arg
```

Differential Revision: D69998314

davidberard98 reviewed (approved):
LGTM as long as tests / lint passes!
